enhancements/upgrade-bundle.md

```diff
@@ -0,0 +1,260 @@
+to the `run` command. This new command, `run bundle-upgrade`, will perform the upgrade of an operator from a previous version to a new specified version.
+
+Currently, the run command has different subcommands: 
+
+ * `run local`
+ * `run packagemanifests`
+ * `run bundle` - currently being implemented, design document is covered by [run bundle enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/run-bundle.md)
+
+For doing an `bundle-upgrade`, there could be two scenarios: 
+
+* Case 1 - A previous version of the operator is already running on the cluster, and the index image contains this operator version. Now, we we want to upgrade the in cluster running operator to a new version that we specify. 
+
+* Case 2 - No previous version of an operator is running on the cluster, so we provide the new bundle image which is accessible and pullable within the cluster, which contains our desired version of the operator for the upgrade. 
+
+In case 1, we will have to figure out which catalog source the existing operator was installed in and somehow dynamically inject the new bundle image into the index image for OLM to detect this change and upgrade to the new version we provided. 
+
+In this proposal, we will design bundle upgrade for case 2. For this case, we also assume that an index image already exists that has the from version of the operator bundle. So we don't create a new index image. We use the existing index image (which is mandatory to provide) for the operator and provide the new bundle image for the upgrade. For this to happen, we need to specify starting CSV.
+Generally, when there is no starting CSV specified, OLM always detects and installs the latest version available for an operator in a particular  channel. For example, in production case, if an operator is running and there's a subscription for it, and there comes a fix available in a stable channel, OLM will detect and upgrade by default. 
+
+Let's consider a few scenarios: 
+
+  * Verifying operator upgrades
+    * To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully upgraded. We will ensure that the CSV has status `Installed` in the appropriate namespaces.  
+
+  * Handling subscription failures 
+    * If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. We then log a message that "run bundle-upgrade was not successful", and exit. 
```

Comment:
What happens if there are multiple versions in the chain between the InstalledCSV and CurrentCSV? OLM's behavior dictates that we we step through each upgrade. The state described here is transient while upgrades are progressing, so I wouldn't outright fail the upgrade based on this condition.